cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -321,7 +323,7 @@ protected void rebalanceForBroker(Broker broker,
         // return if the broker is already within limits.
         return;
       }
-      moveImmigrantsOnly = !clusterModel.deadBrokers().isEmpty();
+      moveImmigrantsOnly = !clusterModel.deadBrokers().isEmpty() || optimizationOptions.onlyMoveImmigrantReplicas();
```

Comment:
I am not sure if our assumption holds: how did we conclude that an overloaded broker (e.g. in terms of disk utilization) would never have immigrant replicas when there is a dead broker?

Let's say I have 2 goals
1. `RackAwareGoal`
2. `DiskUsageDistributionGoal`
3 brokers: `B0` dead, `B1`, `B2` alive
let's assume `B2` has no replicas and receives no replicas during optimization of `RackAwareGoal`
-- i.e. `B1` has 1 immigrant replica and 9 non-immigrant replica with high load
how does `B1` stop giving away non-immigrant replicas to `B2`?

Is it possible to add a test case (if missing) for each goal to ensure that they handle it as expected?

Reply:
It is a valid concern and I am able to reproduce this issue locally with a deterministic test.
There is actually a blind spot in our current self-healing test. I have a patch to fix this issue and also add more test coverage in PR #810